src/internal/storage/track/postgres_tracker.go

```diff
@@ -20,48 +20,89 @@ func NewPostgresTracker(db *sqlx.DB) Tracker {
 	return &postgresTracker{db: db}
 }
 
-func (t *postgresTracker) CreateObject(ctx context.Context, id string, pointsTo []string, ttl time.Duration) error {
+func (t *postgresTracker) DB() *sqlx.DB {
+	return t.db
+}
+
+func (t *postgresTracker) CreateTx(tx *sqlx.Tx, id string, pointsTo []string, ttl time.Duration) error {
 	for _, dwn := range pointsTo {
 		if dwn == id {
 			return ErrSelfReference
 		}
 	}
-	return t.withTx(ctx, func(tx *sqlx.Tx) error {
-		var oid int
-		if err := func() error {
-			if ttl != NoTTL {
-				return tx.GetContext(ctx, &oid,
-					`INSERT INTO storage.tracker_objects (str_id, expires_at)
-				VALUES ($1, CURRENT_TIMESTAMP + $2 * interval '1 microsecond')
-				ON CONFLICT (str_id) DO NOTHING
-				RETURNING int_id
-				`, id, ttl.Microseconds())
-			}
-			return tx.GetContext(ctx, &oid,
-				`INSERT INTO storage.tracker_objects (str_id)
-				VALUES ($1)
-				ON CONFLICT (str_id) DO NOTHING
-				RETURNING int_id
-				`, id)
-		}(); err != nil {
-			if err == sql.ErrNoRows {
-				err = ErrObjectExists
-			}
+	pointsTo = dedupedStrings(pointsTo)
+	// create an object or update the ttl of an existing one
+	intID, created, err := t.putObject(tx, id, ttl)
+	if err != nil {
+		return err
+	}
+	if !created {
+		dwn, err := t.getDownstream(tx, intID)
+		if err != nil {
 			return err
 		}
-		var pointsToInts []int
-		if err := tx.SelectContext(ctx, &pointsToInts,
-			`INSERT INTO storage.tracker_refs (from_id, to_id)
-			SELECT $1, int_id FROM storage.tracker_objects WHERE str_id = ANY($2)
-			RETURNING to_id`,
-			oid, pq.StringArray(pointsTo)); err != nil {
-			return err
+		if !stringsMatch(pointsTo, dwn) {
+			return ErrDifferentObjectExists
+		}
+		return nil
+	}
+	return t.addReferences(tx, intID, pointsTo)
+}
+
+// putObject creates or updates the object at id, to have the max of the current and new ttl.
+// If ttl == NoTTL, then the ttl is removed.
+func (t *postgresTracker) putObject(tx *sqlx.Tx, id string, ttl time.Duration) (int, bool, error) {
+	// About xmax https://stackoverflow.com/a/39204667
+	res := struct {
+		IntID int `db:"int_id"`
+		XMax  int `db:"xmax"`
+	}{}
+	if ttl != NoTTL {
+		if err := tx.Get(&res,
+			`INSERT INTO storage.tracker_objects (str_id, expires_at)
+			VALUES ($1, CURRENT_TIMESTAMP + $2 * interval '1 microsecond')
+			ON CONFLICT (str_id) DO
+				UPDATE SET expires_at = greatest(
```

Comment:
Do we not want to support decreasing the ttl?

Reply:
You can call SetTTL to do that.  If two callers create a chunk, we can't change expires at to worse than what either asked for.